src/compiler/crystal/semantic/call_error.cr

```diff
@@ -194,6 +176,320 @@ class Crystal::Call
+
+    named_args_types.try &.each do |named_arg|
+      found_index = a_def.args.index { |arg| arg.external_name == named_arg.name }
+      if found_index
+        min_size = arg_types.size
+        if found_index < min_size
+          arguments_already_specified << named_arg.name
+        end
+      elsif !a_def.double_splat
+        similar_name = Levenshtein.find(named_arg.name, a_def.args.select(&.default_value).map(&.external_name))
+
+        extra_named_argument_names << named_arg.name
+        extra_named_argument_similar_names << similar_name
+      end
+    end
+
+    unless arguments_already_specified.empty?
+      return ArgumentsAlreadySpecified.new(arguments_already_specified)
+    end
+
+    unless extra_named_argument_names.empty?
+      return ExtraNamedArguments.new(extra_named_argument_names, extra_named_argument_similar_names)
+    end
+
+    # For now let's not deal with splats
+    return if a_def.splat_index
+
+    a_def_owner = a_def.owner
+
+    # This is the actual instantiated type where the method was instantiated
+    instantiated_owner = owner
+
+    owner.ancestors.each do |ancestor|
+      if a_def_owner == ancestor
+        instantiated_owner = ancestor
+        break
+      end
+
+      # If the method is defined in a generic uninstantiated type
+      # then the method instantiation happens on the instantiated generic
+      # type whose generic type is that uninstantiated one.
+      if a_def_owner.is_a?(GenericType) &&
+         ancestor.is_a?(GenericInstanceType) &&
+         ancestor.generic_type == a_def_owner
+        instantiated_owner = ancestor
+        break
+      end
+    end
+
+    match_context = MatchContext.new(
+      instantiated_type: instantiated_owner,
+      defining_type: instantiated_owner,
+      def_free_vars: a_def.free_vars,
+    )
+
+    arguments_type_mismatch = [] of ArgumentTypeMismatch
+
+    arg_types.each_with_index do |arg_type, i|
+      def_arg = a_def.args[i]?
+      next unless def_arg
+
+      check_argument_type_mismatch(def_arg, i, arg_type, match_context, arguments_type_mismatch)
+    end
+
+    named_args_types.try &.each do |named_arg|
+      def_arg = a_def.args.find &.external_name.==(named_arg.name)
+      next unless def_arg
+
+      check_argument_type_mismatch(def_arg, named_arg.name, named_arg.type, match_context, arguments_type_mismatch)
+    end
+
+    unless arguments_type_mismatch.empty?
+      return ArgumentsTypeMismatch.new(arguments_type_mismatch)
+    end
+
+    nil
+  end
+
+  private def check_argument_type_mismatch(def_arg, index_or_name, arg_type, match_context, arguments_type_mismatch)
+    restricted = arg_type.restrict(def_arg, match_context)
+    unless restricted
+      expected_type = def_arg.type?
+      unless expected_type
+        restriction = def_arg.restriction
+        if restriction
+          expected_type = match_context.instantiated_type.lookup_type?(restriction, free_vars: match_context.free_vars)
```

Comment:
I think this might have issues with types like `Array(Array)`.

Reply:
Thanks! Before reviewing the code I'd like to know if we all agree with this going forward. The PR is just a proof that it can be done.

Reply:
Given that we are going forward, if you can provide a scenario where the current code doesn't work well, let me know.